FAIL: TestUpdateBlob flake

Seeing this locally, the test fails non-deterministically.
```
--- FAIL: TestUpdateBlob (0.00s)
    --- FAIL: TestUpdateBlob/valid (0.00s)
        controller.go:150: Unexpected call to *mock_storage.MockBlob.CreateBlockBlobFromReader([0xc00039dad0 <nil>]) at /home/kargakis/go/src/github.com/openshift/openshift-azure/pkg/util/mocks/mock_azureclient/mock_storage/storage.go:198 because: 
            Expected call at /home/kargakis/go/src/github.com/openshift/openshift-azure/pkg/cluster/deploy_test.go:190 doesn't match the argument at index 0.
            Got: &{[91 123 34 105 110 115 116 97 110 99 101 78 97 109 101 34 58 34 115 115 45 99 111 109 112 117 116 101 95 48 34 44 34 115 99 97 108 101 115 101 116 72 97 115 104 34 58 34 55 120 57 57 61 34 125 44 123 34 105 110 115 116 97 110 99 101 78 97 109 101 34 58 34 115 115 45 105 110 102 114 97 95 48 34 44 34 115 99 97 108 101 115 101 116 72 97 115 104 34 58 34 52 53 34 125 93] 0 -1}
            Want: is equal to &{[91 123 34 105 110 115 116 97 110 99 101 78 97 109 101 34 58 34 115 115 45 105 110 102 114 97 95 48 34 44 34 115 99 97 108 101 115 101 116 72 97 115 104 34 58 34 52 53 34 125 44 123 34 105 110 115 116 97 110 99 101 78 97 109 101 34 58 34 115 115 45 99 111 109 112 117 116 101 95 48 34 44 34 115 99 97 108 101 115 101 116 72 97 115 104 34 58 34 55 120 57 57 61 34 125 93] 0 -1}
time="2018-11-27T10:55:20+01:00" level=info msg="drain: node not found, skipping"
time="2018-11-27T10:55:20+01:00" level=info msg="waiting for master-000000 to be ready"
time="2018-11-27T10:55:20+01:00" level=info msg="waiting for master-00000A to be ready"
time="2018-11-27T10:55:20+01:00" level=info msg="waiting for infra-000000 to be ready"
time="2018-11-27T10:55:20+01:00" level=info msg="waiting for compute-000000 to be ready"
time="2018-11-27T10:55:20+01:00" level=info msg="skipping vm \"ss-master_0\" since it's already updated"
time="2018-11-27T10:55:20+01:00" level=info msg="skipping vm \"ss-master_0\" since it's already updated"
time="2018-11-27T10:55:20+01:00" level=info msg="skipping vm \"ss-master_1\" since it's already updated"
time="2018-11-27T10:55:20+01:00" level=info msg="skipping vm \"ss-master_2\" since it's already updated"
FAIL
FAIL	github.com/openshift/openshift-azure/pkg/cluster	0.010s
```
/kind test-flake

You can try reproducing with the following script:
```bash
while true; do go test ./pkg/cluster/ -count 1; if [[ $? != "0" ]]; then break; fi; done
```
@asalkeld

--- a/pkg/cluster/deploy.go
+++ b/pkg/cluster/deploy.go
@@ -1,13 +1,11 @@
 package cluster
 
 import (
-	"bytes"
 	"context"
 	"crypto/sha256"
 	"encoding/base64"
 	"encoding/json"
 	"fmt"
-	"io/ioutil"
 
 	"github.com/openshift/openshift-azure/pkg/api"
 	"github.com/openshift/openshift-azure/pkg/jsonpath"
@@ -46,15 +44,9 @@ func (u *simpleUpgrader) Deploy(ctx context.Context, cs *api.OpenShiftManagedClu
 	return nil
 }
 
-type updateblob map[instanceName]hash
 type scalesetName string
 type instanceName string
 type hash string
-
-type vmInfo struct {
-	InstanceName instanceName `json:"instanceName,omitempty"`
-	ScalesetHash hash         `json:"scalesetHash,omitempty"`
-}
 
 func hashScaleSets(azuretemplate map[string]interface{}) (map[scalesetName]hash, error) {
 	ssHashes := make(map[scalesetName]hash)
@@ -111,42 +103,4 @@ func (u *simpleUpgrader) initializeUpdateBlob(cs *api.OpenShiftManagedCluster, s
 		}
 	}
 	return u.writeUpdateBlob(vmHashes)
-}
-
-func (u *simpleUpgrader) writeUpdateBlob(b updateblob) error {
-	blob := make([]vmInfo, 0, len(b))
-	for instancename, hash := range b {
-		blob = append(blob, vmInfo{
-			InstanceName: instancename,
-			ScalesetHash: hash,
-		})
-	}
-	data, err := json.Marshal(blob)
-	if err != nil {
-		return err
-	}
-	return u.updateBlob.CreateBlockBlobFromReader(bytes.NewReader(data), nil)
-}
-
-func (u *simpleUpgrader) readUpdateBlob() (updateblob, error) {
-	rc, err := u.updateBlob.Get(nil)
-	if err != nil {
-		return nil, err
-	}
-	defer rc.Close()
-
-	data, err := ioutil.ReadAll(rc)
-	if err != nil {
-		return nil, err
-	}
-
-	var blob []vmInfo
-	if err := json.Unmarshal(data, &blob); err != nil {
-		return nil, err
-	}
-	b := updateblob{}
-	for _, vi := range blob {
-		b[vi.InstanceName] = vi.ScalesetHash
-	}
-	return b, nil
 }

--- a/pkg/cluster/deploy.go
+++ b/pkg/cluster/deploy.go
@@ -109,10 +109,10 @@ func (u *simpleUpgrader) initializeUpdateBlob(cs *api.OpenShiftManagedCluster, s
 			vmHashes[name] = ssHashes[scalesetName("ss-"+profile.Name)]
 		}
 	}
-	return u.updateBlob(vmHashes)
+	return u.writeUpdateBlob(vmHashes)
 }
 
-func (u *simpleUpgrader) updateBlob(b map[instanceName]hash) error {
+func (u *simpleUpgrader) writeUpdateBlob(b map[instanceName]hash) error {
 	blob := make([]vmInfo, 0, len(b))
 	for instancename, hash := range b {
 		blob = append(blob, vmInfo{
@@ -130,7 +130,7 @@ func (u *simpleUpgrader) updateBlob(b map[instanceName]hash) error {
 	return bc.CreateBlockBlobFromReader(bytes.NewReader(data), nil)
 }
 
-func (u *simpleUpgrader) readBlob() (map[instanceName]hash, error) {
+func (u *simpleUpgrader) readUpdateBlob() (map[instanceName]hash, error) {
 	bsc := u.storageClient.GetBlobService()
 	c := bsc.GetContainerReference(updateContainerName)
 	bc := c.GetBlobReference(updateBlobName)

--- a/pkg/cluster/deploy.go
+++ b/pkg/cluster/deploy.go
@@ -46,6 +46,7 @@ func (u *simpleUpgrader) Deploy(ctx context.Context, cs *api.OpenShiftManagedClu
 	return nil
 }
 
+type updateblob map[instanceName]hash
 type scalesetName string
 type instanceName string
 type hash string
@@ -102,7 +103,7 @@ func deepCopy(in map[string]interface{}) map[string]interface{} {
 }
 
 func (u *simpleUpgrader) initializeUpdateBlob(cs *api.OpenShiftManagedCluster, ssHashes map[scalesetName]hash) error {
-	vmHashes := make(map[instanceName]hash)
+	vmHashes := updateblob{}
 	for _, profile := range cs.Properties.AgentPoolProfiles {
 		for i := 0; i < profile.Count; i++ {
 			name := instanceName(fmt.Sprintf("ss-%s_%d", profile.Name, i))
@@ -112,7 +113,7 @@ func (u *simpleUpgrader) initializeUpdateBlob(cs *api.OpenShiftManagedCluster, s
 	return u.writeUpdateBlob(vmHashes)
 }
 
-func (u *simpleUpgrader) writeUpdateBlob(b map[instanceName]hash) error {
+func (u *simpleUpgrader) writeUpdateBlob(b updateblob) error {
 	blob := make([]vmInfo, 0, len(b))
 	for instancename, hash := range b {
 		blob = append(blob, vmInfo{
@@ -127,7 +128,7 @@ func (u *simpleUpgrader) writeUpdateBlob(b map[instanceName]hash) error {
 	return u.updateBlob.CreateBlockBlobFromReader(bytes.NewReader(data), nil)
 }
 
-func (u *simpleUpgrader) readUpdateBlob() (map[instanceName]hash, error) {
+func (u *simpleUpgrader) readUpdateBlob() (updateblob, error) {
 	rc, err := u.updateBlob.Get(nil)
 	if err != nil {
 		return nil, err
@@ -143,7 +144,7 @@ func (u *simpleUpgrader) readUpdateBlob() (map[instanceName]hash, error) {
 	if err := json.Unmarshal(data, &blob); err != nil {
 		return nil, err
 	}
-	b := make(map[instanceName]hash)
+	b := updateblob{}
 	for _, vi := range blob {
 		b[vi.InstanceName] = vi.ScalesetHash
 	}

--- a/pkg/cluster/deploy.go
+++ b/pkg/cluster/deploy.go
@@ -124,18 +124,11 @@ func (u *simpleUpgrader) writeUpdateBlob(b map[instanceName]hash) error {
 	if err != nil {
 		return err
 	}
-	bsc := u.storageClient.GetBlobService()
-	c := bsc.GetContainerReference(updateContainerName)
-	bc := c.GetBlobReference(updateBlobName)
-	return bc.CreateBlockBlobFromReader(bytes.NewReader(data), nil)
+	return u.updateBlob.CreateBlockBlobFromReader(bytes.NewReader(data), nil)
 }
 
 func (u *simpleUpgrader) readUpdateBlob() (map[instanceName]hash, error) {
-	bsc := u.storageClient.GetBlobService()
-	c := bsc.GetContainerReference(updateContainerName)
-	bc := c.GetBlobReference(updateBlobName)
-
-	rc, err := bc.Get(nil)
+	rc, err := u.updateBlob.Get(nil)
 	if err != nil {
 		return nil, err
 	}

--- a/pkg/cluster/deploy.go
+++ b/pkg/cluster/deploy.go
@@ -112,9 +112,6 @@ func (u *simpleUpgrader) initializeUpdateBlob(cs *api.OpenShiftManagedCluster, s
 	return u.updateBlob(vmHashes)
 }
 
-const updateContainerName = "update"
-const updateBlobName = "update"
-
 func (u *simpleUpgrader) updateBlob(b map[instanceName]hash) error {
 	blob := make([]vmInfo, 0, len(b))
 	for instancename, hash := range b {

--- a/pkg/cluster/deploy.go
+++ b/pkg/cluster/deploy.go
@@ -95,12 +95,12 @@ func deepCopy(in map[string]interface{}) map[string]interface{} {
 }
 
 func (u *simpleUpgrader) initializeUpdateBlob(cs *api.OpenShiftManagedCluster, ssHashes map[scalesetName]hash) error {
-	vmHashes := updateblob{}
+	blob := updateblob{}
 	for _, profile := range cs.Properties.AgentPoolProfiles {
 		for i := 0; i < profile.Count; i++ {
 			name := instanceName(fmt.Sprintf("ss-%s_%d", profile.Name, i))
-			vmHashes[name] = ssHashes[scalesetName("ss-"+profile.Name)]
+			blob[name] = ssHashes[scalesetName("ss-"+profile.Name)]
 		}
 	}
-	return u.writeUpdateBlob(vmHashes)
+	return u.writeUpdateBlob(blob)
 }

--- a/pkg/cluster/deploy_test.go
+++ b/pkg/cluster/deploy_test.go
@@ -125,21 +125,15 @@ func TestReadUpdateBlob(t *testing.T) {
 		},
 	}
 	gmc := gomock.NewController(t)
-	storageClient := mock_storage.NewMockClient(gmc)
 	u := &simpleUpgrader{
-		pluginConfig:  api.PluginConfig{},
-		storageClient: storageClient,
+		pluginConfig: api.PluginConfig{},
 	}
 
 	for _, tt := range tests {
-		bsa := mock_storage.NewMockBlobStorageClient(gmc)
-		storageClient.EXPECT().GetBlobService().Return(bsa)
-		updateCr := mock_storage.NewMockContainer(gmc)
-		bsa.EXPECT().GetContainerReference(updateContainerName).Return(updateCr)
 		updateBlob := mock_storage.NewMockBlob(gmc)
-		updateCr.EXPECT().GetBlobReference(updateBlobName).Return(updateBlob)
 		data := ioutil.NopCloser(strings.NewReader(tt.blob))
 		updateBlob.EXPECT().Get(nil).Return(data, nil)
+		u.updateBlob = updateBlob
 
 		got, err := u.readUpdateBlob()
 		if (err != nil) != (len(tt.wantErr) > 0) {
@@ -179,18 +173,12 @@ func TestWriteUpdateBlob(t *testing.T) {
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 			gmc := gomock.NewController(t)
-			storageClient := mock_storage.NewMockClient(gmc)
 
-			bsa := mock_storage.NewMockBlobStorageClient(gmc)
-			storageClient.EXPECT().GetBlobService().Return(bsa)
-			updateCr := mock_storage.NewMockContainer(gmc)
-			bsa.EXPECT().GetContainerReference(updateContainerName).Return(updateCr)
 			updateBlob := mock_storage.NewMockBlob(gmc)
-			updateCr.EXPECT().GetBlobReference(updateBlobName).Return(updateBlob)
 			updateBlob.EXPECT().CreateBlockBlobFromReader(bytes.NewReader([]byte(tt.blob)), nil)
 			u := &simpleUpgrader{
-				pluginConfig:  api.PluginConfig{},
-				storageClient: storageClient,
+				pluginConfig: api.PluginConfig{},
+				updateBlob:   updateBlob,
 			}
 			if err := u.writeUpdateBlob(tt.b); (err != nil) != (tt.wantErr != "") {
 				t.Errorf("simpleUpgrader.writeUpdateBlob() error = %v, wantErr %v", err, tt.wantErr)

--- a/pkg/cluster/deploy_test.go
+++ b/pkg/cluster/deploy_test.go
@@ -102,7 +102,7 @@ func TestHashScaleSets(t *testing.T) {
 	}
 }
 
-func TestReadBlob(t *testing.T) {
+func TestReadUpdateBlob(t *testing.T) {
 	tests := []struct {
 		name    string
 		want    map[instanceName]hash
@@ -141,21 +141,21 @@ func TestReadBlob(t *testing.T) {
 		data := ioutil.NopCloser(strings.NewReader(tt.blob))
 		updateBlob.EXPECT().Get(nil).Return(data, nil)
 
-		got, err := u.readBlob()
+		got, err := u.readUpdateBlob()
 		if (err != nil) != (len(tt.wantErr) > 0) {
-			t.Errorf("simpleUpgrader.readBlob() error = %v, wantErr %v", err, tt.wantErr)
+			t.Errorf("simpleUpgrader.readUpdateBlob() error = %v, wantErr %v", err, tt.wantErr)
 			return
 		}
 		if len(tt.wantErr) > 0 && !strings.Contains(err.Error(), tt.wantErr) {
-			t.Errorf("simpleUpgrader.readBlob() error = %v, wantErr %v", err, tt.wantErr)
+			t.Errorf("simpleUpgrader.readUpdateBlob() error = %v, wantErr %v", err, tt.wantErr)
 		}
 		if !reflect.DeepEqual(got, tt.want) && len(tt.wantErr) == 0 {
-			t.Errorf("simpleUpgrader.readBlob() = %v, want %v", got, tt.want)
+			t.Errorf("simpleUpgrader.readUpdateBlob() = %v, want %v", got, tt.want)
 		}
 	}
 }
 
-func TestUpdateBlob(t *testing.T) {
+func TestWriteUpdateBlob(t *testing.T) {
 	tests := []struct {
 		name    string
 		b       map[instanceName]hash
@@ -192,8 +192,8 @@ func TestUpdateBlob(t *testing.T) {
 				pluginConfig:  api.PluginConfig{},
 				storageClient: storageClient,
 			}
-			if err := u.updateBlob(tt.b); (err != nil) != (tt.wantErr != "") {
-				t.Errorf("simpleUpgrader.updateBlob() error = %v, wantErr %v", err, tt.wantErr)
+			if err := u.writeUpdateBlob(tt.b); (err != nil) != (tt.wantErr != "") {
+				t.Errorf("simpleUpgrader.writeUpdateBlob() error = %v, wantErr %v", err, tt.wantErr)
 			}
 		})
 	}

--- a/pkg/cluster/deploy_test.go
+++ b/pkg/cluster/deploy_test.go
@@ -9,7 +9,6 @@ import (
 
 	gomock "github.com/golang/mock/gomock"
 
-	"github.com/openshift/openshift-azure/pkg/api"
 	"github.com/openshift/openshift-azure/pkg/util/mocks/mock_azureclient/mock_storage"
 )
 
@@ -111,8 +110,6 @@ func TestReadUpdateBlob(t *testing.T) {
 	}{
 		{
 			name:    "empty",
-			want:    map[instanceName]hash{},
-			blob:    "",
 			wantErr: "unexpected end of JSON input",
 		},
 		{
@@ -121,19 +118,17 @@ func TestReadUpdateBlob(t *testing.T) {
 				"ss-infra_0":   "45",
 				"ss-compute_0": "7x99=",
 			},
-			blob: "[{\"instanceName\": \"ss-infra_0\", \"scalesetHash\": \"45\"},{\"instanceName\":\"ss-compute_0\",\"scalesetHash\":\"7x99=\"}]",
+			blob: `[{"instanceName": "ss-infra_0", "scalesetHash": "45"},{"instanceName":"ss-compute_0","scalesetHash":"7x99="}]`,
 		},
 	}
 	gmc := gomock.NewController(t)
-	u := &simpleUpgrader{
-		pluginConfig: api.PluginConfig{},
-	}
-
 	for _, tt := range tests {
 		updateBlob := mock_storage.NewMockBlob(gmc)
 		data := ioutil.NopCloser(strings.NewReader(tt.blob))
 		updateBlob.EXPECT().Get(nil).Return(data, nil)
-		u.updateBlob = updateBlob
+		u := &simpleUpgrader{
+			updateBlob: updateBlob,
+		}
 
 		got, err := u.readUpdateBlob()
 		if (err != nil) != (len(tt.wantErr) > 0) {
@@ -158,7 +153,6 @@ func TestWriteUpdateBlob(t *testing.T) {
 	}{
 		{
 			name: "empty",
-			b:    map[instanceName]hash{},
 			blob: "[]",
 		},
 		{
@@ -167,22 +161,19 @@ func TestWriteUpdateBlob(t *testing.T) {
 				"ss-infra_0":   "45",
 				"ss-compute_0": "7x99=",
 			},
-			blob: "[{\"instanceName\":\"ss-infra_0\",\"scalesetHash\":\"45\"},{\"instanceName\":\"ss-compute_0\",\"scalesetHash\":\"7x99=\"}]",
+			blob: `[{"instanceName":"ss-infra_0","scalesetHash":"45"},{"instanceName":"ss-compute_0","scalesetHash":"7x99="}]`,
 		},
 	}
+	gmc := gomock.NewController(t)
 	for _, tt := range tests {
-		t.Run(tt.name, func(t *testing.T) {
-			gmc := gomock.NewController(t)
+		updateBlob := mock_storage.NewMockBlob(gmc)
+		updateBlob.EXPECT().CreateBlockBlobFromReader(bytes.NewReader([]byte(tt.blob)), nil)
+		u := &simpleUpgrader{
+			updateBlob: updateBlob,
+		}
 
-			updateBlob := mock_storage.NewMockBlob(gmc)
-			updateBlob.EXPECT().CreateBlockBlobFromReader(bytes.NewReader([]byte(tt.blob)), nil)
-			u := &simpleUpgrader{
-				pluginConfig: api.PluginConfig{},
-				updateBlob:   updateBlob,
-			}
-			if err := u.writeUpdateBlob(tt.b); (err != nil) != (tt.wantErr != "") {
-				t.Errorf("simpleUpgrader.writeUpdateBlob() error = %v, wantErr %v", err, tt.wantErr)
-			}
-		})
+		if err := u.writeUpdateBlob(tt.b); (err != nil) != (tt.wantErr != "") {
+			t.Errorf("simpleUpgrader.writeUpdateBlob() error = %v, wantErr %v", err, tt.wantErr)
+		}
 	}
 }

--- a/pkg/cluster/deploy_test.go
+++ b/pkg/cluster/deploy_test.go
@@ -135,9 +135,9 @@ func TestReadBlob(t *testing.T) {
 		bsa := mock_storage.NewMockBlobStorageClient(gmc)
 		storageClient.EXPECT().GetBlobService().Return(bsa)
 		updateCr := mock_storage.NewMockContainer(gmc)
-		bsa.EXPECT().GetContainerReference("update").Return(updateCr)
+		bsa.EXPECT().GetContainerReference(updateContainerName).Return(updateCr)
 		updateBlob := mock_storage.NewMockBlob(gmc)
-		updateCr.EXPECT().GetBlobReference("update").Return(updateBlob)
+		updateCr.EXPECT().GetBlobReference(updateBlobName).Return(updateBlob)
 		data := ioutil.NopCloser(strings.NewReader(tt.blob))
 		updateBlob.EXPECT().Get(nil).Return(data, nil)
 
@@ -184,9 +184,9 @@ func TestUpdateBlob(t *testing.T) {
 			bsa := mock_storage.NewMockBlobStorageClient(gmc)
 			storageClient.EXPECT().GetBlobService().Return(bsa)
 			updateCr := mock_storage.NewMockContainer(gmc)
-			bsa.EXPECT().GetContainerReference("update").Return(updateCr)
+			bsa.EXPECT().GetContainerReference(updateContainerName).Return(updateCr)
 			updateBlob := mock_storage.NewMockBlob(gmc)
-			updateCr.EXPECT().GetBlobReference("update").Return(updateBlob)
+			updateCr.EXPECT().GetBlobReference(updateBlobName).Return(updateBlob)
 			updateBlob.EXPECT().CreateBlockBlobFromReader(bytes.NewReader([]byte(tt.blob)), nil)
 			u := &simpleUpgrader{
 				pluginConfig:  api.PluginConfig{},

--- a/pkg/cluster/deploy_test.go
+++ b/pkg/cluster/deploy_test.go
@@ -104,7 +104,7 @@ func TestHashScaleSets(t *testing.T) {
 func TestReadUpdateBlob(t *testing.T) {
 	tests := []struct {
 		name    string
-		want    map[instanceName]hash
+		want    updateblob
 		wantErr string
 		blob    string
 	}{
@@ -114,7 +114,7 @@ func TestReadUpdateBlob(t *testing.T) {
 		},
 		{
 			name: "ok",
-			want: map[instanceName]hash{
+			want: updateblob{
 				"ss-infra_0":   "45",
 				"ss-compute_0": "7x99=",
 			},
@@ -147,7 +147,7 @@ func TestReadUpdateBlob(t *testing.T) {
 func TestWriteUpdateBlob(t *testing.T) {
 	tests := []struct {
 		name    string
-		b       map[instanceName]hash
+		b       updateblob
 		wantErr string
 		blob    string
 	}{
@@ -157,7 +157,7 @@ func TestWriteUpdateBlob(t *testing.T) {
 		},
 		{
 			name: "valid",
-			b: map[instanceName]hash{
+			b: updateblob{
 				"ss-infra_0":   "45",
 				"ss-compute_0": "7x99=",
 			},

--- a/pkg/cluster/deploy_test.go
+++ b/pkg/cluster/deploy_test.go
@@ -1,15 +1,8 @@
 package cluster
 
 import (
-	"bytes"
-	"io/ioutil"
 	"reflect"
-	"strings"
 	"testing"
-
-	gomock "github.com/golang/mock/gomock"
-
-	"github.com/openshift/openshift-azure/pkg/util/mocks/mock_azureclient/mock_storage"
 )
 
 const masterHash = "RmID82LhPjuQbCEdiVa5cGCVEkdLGD6iU6ozX3vxkD0="
@@ -97,83 +90,6 @@ func TestHashScaleSets(t *testing.T) {
 		}
 		if !reflect.DeepEqual(got, test.exp) {
 			t.Errorf("%s: expected:\n%#v\ngot:\n%#v", test.name, test.exp, got)
-		}
-	}
-}
-
-func TestReadUpdateBlob(t *testing.T) {
-	tests := []struct {
-		name    string
-		want    updateblob
-		wantErr string
-		blob    string
-	}{
-		{
-			name:    "empty",
-			wantErr: "unexpected end of JSON input",
-		},
-		{
-			name: "ok",
-			want: updateblob{
-				"ss-infra_0":   "45",
-				"ss-compute_0": "7x99=",
-			},
-			blob: `[{"instanceName": "ss-infra_0", "scalesetHash": "45"},{"instanceName":"ss-compute_0","scalesetHash":"7x99="}]`,
-		},
-	}
-	gmc := gomock.NewController(t)
-	for _, tt := range tests {
-		updateBlob := mock_storage.NewMockBlob(gmc)
-		data := ioutil.NopCloser(strings.NewReader(tt.blob))
-		updateBlob.EXPECT().Get(nil).Return(data, nil)
-		u := &simpleUpgrader{
-			updateBlob: updateBlob,
-		}
-
-		got, err := u.readUpdateBlob()
-		if (err != nil) != (len(tt.wantErr) > 0) {
-			t.Errorf("simpleUpgrader.readUpdateBlob() error = %v, wantErr %v", err, tt.wantErr)
-			return
-		}
-		if len(tt.wantErr) > 0 && !strings.Contains(err.Error(), tt.wantErr) {
-			t.Errorf("simpleUpgrader.readUpdateBlob() error = %v, wantErr %v", err, tt.wantErr)
-		}
-		if !reflect.DeepEqual(got, tt.want) && len(tt.wantErr) == 0 {
-			t.Errorf("simpleUpgrader.readUpdateBlob() = %v, want %v", got, tt.want)
-		}
-	}
-}
-
-func TestWriteUpdateBlob(t *testing.T) {
-	tests := []struct {
-		name    string
-		b       updateblob
-		wantErr string
-		blob    string
-	}{
-		{
-			name: "empty",
-			blob: "[]",
-		},
-		{
-			name: "valid",
-			b: updateblob{
-				"ss-infra_0":   "45",
-				"ss-compute_0": "7x99=",
-			},
-			blob: `[{"instanceName":"ss-infra_0","scalesetHash":"45"},{"instanceName":"ss-compute_0","scalesetHash":"7x99="}]`,
-		},
-	}
-	gmc := gomock.NewController(t)
-	for _, tt := range tests {
-		updateBlob := mock_storage.NewMockBlob(gmc)
-		updateBlob.EXPECT().CreateBlockBlobFromReader(bytes.NewReader([]byte(tt.blob)), nil)
-		u := &simpleUpgrader{
-			updateBlob: updateBlob,
-		}
-
-		if err := u.writeUpdateBlob(tt.b); (err != nil) != (tt.wantErr != "") {
-			t.Errorf("simpleUpgrader.writeUpdateBlob() error = %v, wantErr %v", err, tt.wantErr)
 		}
 	}
 }

--- a/pkg/cluster/initialize.go
+++ b/pkg/cluster/initialize.go
@@ -30,7 +30,7 @@ func (u *simpleUpgrader) initialize(ctx context.Context, cs *api.OpenShiftManage
 	}
 
 	// update tracking container
-	c = bsc.GetContainerReference("update")
+	c = bsc.GetContainerReference(updateContainerName)
 	_, err = c.CreateIfNotExists(nil)
 	if err != nil {
 		return err

--- a/pkg/cluster/initialize.go
+++ b/pkg/cluster/initialize.go
@@ -35,6 +35,7 @@ func (u *simpleUpgrader) initialize(ctx context.Context, cs *api.OpenShiftManage
 	if err != nil {
 		return err
 	}
+	u.updateBlob = c.GetBlobReference(updateBlobName)
 
 	// cluster config container
 	c = bsc.GetContainerReference(ConfigContainerName)

--- a/pkg/cluster/initialize_test.go
+++ b/pkg/cluster/initialize_test.go
@@ -6,7 +6,7 @@ import (
 	"encoding/json"
 	"testing"
 
-	gomock "github.com/golang/mock/gomock"
+	"github.com/golang/mock/gomock"
 
 	"github.com/openshift/openshift-azure/pkg/api"
 	"github.com/openshift/openshift-azure/pkg/util/mocks/mock_azureclient/mock_storage"

--- a/pkg/cluster/initialize_test.go
+++ b/pkg/cluster/initialize_test.go
@@ -32,6 +32,7 @@ func TestInitialize(t *testing.T) {
 	updateCr := mock_storage.NewMockContainer(gmc)
 	bsa.EXPECT().GetContainerReference(updateContainerName).Return(updateCr)
 	updateCr.EXPECT().CreateIfNotExists(nil).Return(true, nil)
+	updateCr.EXPECT().GetBlobReference(updateBlobName)
 
 	configCr := mock_storage.NewMockContainer(gmc)
 	bsa.EXPECT().GetContainerReference(ConfigContainerName).Return(configCr)

--- a/pkg/cluster/initialize_test.go
+++ b/pkg/cluster/initialize_test.go
@@ -9,18 +9,14 @@ import (
 	gomock "github.com/golang/mock/gomock"
 
 	"github.com/openshift/openshift-azure/pkg/api"
-	"github.com/openshift/openshift-azure/pkg/util/mocks/mock_azureclient"
 	"github.com/openshift/openshift-azure/pkg/util/mocks/mock_azureclient/mock_storage"
 )
 
 func TestInitialize(t *testing.T) {
 	gmc := gomock.NewController(t)
-	accountsClient := mock_azureclient.NewMockAccountsClient(gmc)
 	storageClient := mock_storage.NewMockClient(gmc)
-	si := &simpleUpgrader{
-		pluginConfig:   api.PluginConfig{},
-		accountsClient: accountsClient,
-		storageClient:  storageClient,
+	u := &simpleUpgrader{
+		storageClient: storageClient,
 	}
 	bsa := mock_storage.NewMockBlobStorageClient(gmc)
 	storageClient.EXPECT().GetBlobService().Return(bsa)
@@ -49,7 +45,7 @@ func TestInitialize(t *testing.T) {
 	var finalError error
 	configBlob.EXPECT().CreateBlockBlobFromReader(bytes.NewReader(csj), nil).Return(finalError)
 
-	if err := si.initialize(context.Background(), cs); err != finalError {
+	if err := u.initialize(context.Background(), cs); err != finalError {
 		t.Errorf("simpleUpgrader.initialize() error = %v", err)
 	}
 }

--- a/pkg/cluster/initialize_test.go
+++ b/pkg/cluster/initialize_test.go
@@ -30,7 +30,7 @@ func TestInitialize(t *testing.T) {
 	etcdCr.EXPECT().CreateIfNotExists(nil).Return(true, nil)
 
 	updateCr := mock_storage.NewMockContainer(gmc)
-	bsa.EXPECT().GetContainerReference("update").Return(updateCr)
+	bsa.EXPECT().GetContainerReference(updateContainerName).Return(updateCr)
 	updateCr.EXPECT().CreateIfNotExists(nil).Return(true, nil)
 
 	configCr := mock_storage.NewMockContainer(gmc)

--- a/pkg/cluster/ready_test.go
+++ b/pkg/cluster/ready_test.go
@@ -7,7 +7,7 @@ import (
 
 	compute "github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2018-06-01/compute"
 	"github.com/Azure/go-autorest/autorest/to"
-	gomock "github.com/golang/mock/gomock"
+	"github.com/golang/mock/gomock"
 	"github.com/sirupsen/logrus"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

--- a/pkg/cluster/types.go
+++ b/pkg/cluster/types.go
@@ -39,6 +39,7 @@ type simpleUpgrader struct {
 	pluginConfig   api.PluginConfig
 	accountsClient azureclient.AccountsClient
 	storageClient  storage.Client
+	updateBlob     storage.Blob
 	vmc            azureclient.VirtualMachineScaleSetVMsClient
 	ssc            azureclient.VirtualMachineScaleSetsClient
 	kubeclient     kubernetes.Interface

--- a/pkg/cluster/types.go
+++ b/pkg/cluster/types.go
@@ -21,8 +21,10 @@ import (
 // here follow well known container and blob names
 const (
 	ConfigContainerName     = "config"
-	EtcdBackupContainerName = "etcd"
 	ConfigBlobName          = "config"
+	updateContainerName     = "update"
+	updateBlobName          = "update"
+	EtcdBackupContainerName = "etcd"
 )
 
 // Upgrader is the public interface to the upgrade module used by the plugin.

--- a/pkg/cluster/update.go
+++ b/pkg/cluster/update.go
@@ -81,7 +81,7 @@ func (u *simpleUpgrader) getNodesAndDrain(ctx context.Context, cs *api.OpenShift
 }
 
 func (u *simpleUpgrader) waitForNewNodes(ctx context.Context, cs *api.OpenShiftManagedCluster, nodes map[computerName]struct{}, ssHashes map[scalesetName]hash) error {
-	blob, err := u.readBlob()
+	blob, err := u.readUpdateBlob()
 	if err != nil {
 		return err
 	}
@@ -103,7 +103,7 @@ func (u *simpleUpgrader) waitForNewNodes(ctx context.Context, cs *api.OpenShiftM
 					return err
 				}
 				blob[instanceName(*vm.Name)] = ssHashes[ssNameForVM(&vm)]
-				if err := u.updateBlob(blob); err != nil {
+				if err := u.writeUpdateBlob(blob); err != nil {
 					return err
 				}
 			}
@@ -121,7 +121,7 @@ func (u *simpleUpgrader) waitForNewNodes(ctx context.Context, cs *api.OpenShiftM
 		}
 	}
 	if needsUpdate {
-		return u.updateBlob(blob)
+		return u.writeUpdateBlob(blob)
 	}
 	return nil
 }
@@ -167,7 +167,7 @@ func (u *simpleUpgrader) updatePlusOne(ctx context.Context, cs *api.OpenShiftMan
 		return &api.PluginError{Err: err, Step: api.PluginStepUpdatePlusOneListVMs}
 	}
 
-	blob, err := u.readBlob()
+	blob, err := u.readUpdateBlob()
 	if err != nil {
 		return &api.PluginError{Err: err, Step: api.PluginStepUpdatePlusOneReadBlob}
 	}
@@ -213,7 +213,7 @@ func (u *simpleUpgrader) updatePlusOne(ctx context.Context, cs *api.OpenShiftMan
 				}
 				vmsBefore[*updated.InstanceID] = struct{}{}
 				blob[instanceName(*updated.Name)] = ssHashes[ssNameForVM(&updated)]
-				if err := u.updateBlob(blob); err != nil {
+				if err := u.writeUpdateBlob(blob); err != nil {
 					return &api.PluginError{Err: err, Step: api.PluginStepUpdatePlusOneUpdateBlob}
 				}
 			}
@@ -223,7 +223,7 @@ func (u *simpleUpgrader) updatePlusOne(ctx context.Context, cs *api.OpenShiftMan
 			return &api.PluginError{Err: err, Step: api.PluginStepUpdatePlusOneDeleteVMs}
 		}
 		delete(blob, instanceName(*vm.Name))
-		if err := u.updateBlob(blob); err != nil {
+		if err := u.writeUpdateBlob(blob); err != nil {
 			return &api.PluginError{Err: err, Step: api.PluginStepUpdatePlusOneUpdateBlob}
 		}
 	}
@@ -260,7 +260,7 @@ func (u *simpleUpgrader) updateInPlace(ctx context.Context, cs *api.OpenShiftMan
 		return &api.PluginError{Err: err, Step: api.PluginStepUpdateInPlaceSortMasters}
 	}
 
-	blob, err := u.readBlob()
+	blob, err := u.readUpdateBlob()
 	if err != nil {
 		return &api.PluginError{Err: err, Step: api.PluginStepUpdateInPlaceReadBlob}
 	}
@@ -335,7 +335,7 @@ func (u *simpleUpgrader) updateInPlace(ctx context.Context, cs *api.OpenShiftMan
 		}
 
 		blob[instanceName(*vm.Name)] = ssHashes[ssNameForVM(&vm)]
-		if err := u.updateBlob(blob); err != nil {
+		if err := u.writeUpdateBlob(blob); err != nil {
 			return &api.PluginError{Err: err, Step: api.PluginStepUpdateInPlaceUpdateBlob}
 		}
 	}

--- a/pkg/cluster/update.go
+++ b/pkg/cluster/update.go
@@ -231,7 +231,7 @@ func (u *simpleUpgrader) updatePlusOne(ctx context.Context, cs *api.OpenShiftMan
 	return nil
 }
 
-func (u *simpleUpgrader) filterOldVMs(vms []compute.VirtualMachineScaleSetVM, blob map[instanceName]hash, ssHashes map[scalesetName]hash) []compute.VirtualMachineScaleSetVM {
+func (u *simpleUpgrader) filterOldVMs(vms []compute.VirtualMachineScaleSetVM, blob updateblob, ssHashes map[scalesetName]hash) []compute.VirtualMachineScaleSetVM {
 	var oldVMs []compute.VirtualMachineScaleSetVM
 	for _, vm := range vms {
 		if blob[instanceName(*vm.Name)] != ssHashes[ssNameForVM(&vm)] {

--- a/pkg/cluster/update_test.go
+++ b/pkg/cluster/update_test.go
@@ -13,7 +13,7 @@ func TestFilterOldVMs(t *testing.T) {
 	tests := []struct {
 		name     string
 		vms      []compute.VirtualMachineScaleSetVM
-		blob     map[instanceName]hash
+		blob     updateblob
 		ssHashes map[scalesetName]hash
 		exp      []compute.VirtualMachineScaleSetVM
 	}{
@@ -30,7 +30,7 @@ func TestFilterOldVMs(t *testing.T) {
 					Name: to.StringPtr("ss-master_2"),
 				},
 			},
-			blob: map[instanceName]hash{
+			blob: updateblob{
 				"ss-master_0": "newhash",
 				"ss-master_1": "oldhash",
 				"ss-master_2": "oldhash",
@@ -60,7 +60,7 @@ func TestFilterOldVMs(t *testing.T) {
 					Name: to.StringPtr("ss-master_2"),
 				},
 			},
-			blob: map[instanceName]hash{
+			blob: updateblob{
 				"ss-master_0": "newhash",
 				"ss-master_1": "newhash",
 				"ss-master_2": "newhash",

--- a/pkg/cluster/updateblob.go
+++ b/pkg/cluster/updateblob.go
@@ -3,7 +3,6 @@ package cluster
 import (
 	"bytes"
 	"encoding/json"
-	"io/ioutil"
 )
 
 type updateblob map[instanceName]hash
@@ -57,14 +56,12 @@ func (u *simpleUpgrader) readUpdateBlob() (updateblob, error) {
 	}
 	defer rc.Close()
 
-	data, err := ioutil.ReadAll(rc)
-	if err != nil {
+	d := json.NewDecoder(rc)
+
+	b := updateblob{}
+	if err := d.Decode(&b); err != nil {
 		return nil, err
 	}
 
-	b := updateblob{}
-	if err := json.Unmarshal(data, &b); err != nil {
-		return nil, err
-	}
 	return b, nil
 }

--- a/pkg/cluster/updateblob.go
+++ b/pkg/cluster/updateblob.go
@@ -1,0 +1,52 @@
+package cluster
+
+import (
+	"bytes"
+	"encoding/json"
+	"io/ioutil"
+)
+
+type updateblob map[instanceName]hash
+
+type vmInfo struct {
+	InstanceName instanceName `json:"instanceName,omitempty"`
+	ScalesetHash hash         `json:"scalesetHash,omitempty"`
+}
+
+func (u *simpleUpgrader) writeUpdateBlob(b updateblob) error {
+	blob := make([]vmInfo, 0, len(b))
+	for instancename, hash := range b {
+		blob = append(blob, vmInfo{
+			InstanceName: instancename,
+			ScalesetHash: hash,
+		})
+	}
+	data, err := json.Marshal(blob)
+	if err != nil {
+		return err
+	}
+	return u.updateBlob.CreateBlockBlobFromReader(bytes.NewReader(data), nil)
+}
+
+func (u *simpleUpgrader) readUpdateBlob() (updateblob, error) {
+	rc, err := u.updateBlob.Get(nil)
+	if err != nil {
+		return nil, err
+	}
+	defer rc.Close()
+
+	data, err := ioutil.ReadAll(rc)
+	if err != nil {
+		return nil, err
+	}
+
+	var blob []vmInfo
+	if err := json.Unmarshal(data, &blob); err != nil {
+		return nil, err
+	}
+	b := updateblob{}
+	for _, vi := range blob {
+		b[vi.InstanceName] = vi.ScalesetHash
+	}
+	return b, nil
+}

--- a/pkg/cluster/updateblob.go
+++ b/pkg/cluster/updateblob.go
@@ -8,19 +8,41 @@ import (
 
 type updateblob map[instanceName]hash
 
+var _ json.Marshaler = &updateblob{}
+var _ json.Unmarshaler = &updateblob{}
+
 type vmInfo struct {
 	InstanceName instanceName `json:"instanceName,omitempty"`
 	ScalesetHash hash         `json:"scalesetHash,omitempty"`
 }
 
-func (u *simpleUpgrader) writeUpdateBlob(b updateblob) error {
-	blob := make([]vmInfo, 0, len(b))
-	for instancename, hash := range b {
-		blob = append(blob, vmInfo{
+func (blob updateblob) MarshalJSON() ([]byte, error) {
+	slice := make([]vmInfo, 0, len(blob))
+	for instancename, hash := range blob {
+		slice = append(slice, vmInfo{
 			InstanceName: instancename,
 			ScalesetHash: hash,
 		})
 	}
+
+	return json.Marshal(slice)
+}
+
+func (blob *updateblob) UnmarshalJSON(data []byte) error {
+	var slice []vmInfo
+	if err := json.Unmarshal(data, &slice); err != nil {
+		return err
+	}
+
+	*blob = updateblob{}
+	for _, vi := range slice {
+		(*blob)[vi.InstanceName] = vi.ScalesetHash
+	}
+
+	return nil
+}
+
+func (u *simpleUpgrader) writeUpdateBlob(blob updateblob) error {
 	data, err := json.Marshal(blob)
 	if err != nil {
 		return err
@@ -40,13 +62,9 @@ func (u *simpleUpgrader) readUpdateBlob() (updateblob, error) {
 		return nil, err
 	}
 
-	var blob []vmInfo
-	if err := json.Unmarshal(data, &blob); err != nil {
-		return nil, err
-	}
 	b := updateblob{}
-	for _, vi := range blob {
-		b[vi.InstanceName] = vi.ScalesetHash
+	if err := json.Unmarshal(data, &b); err != nil {
+		return nil, err
 	}
 	return b, nil
 }

--- a/pkg/cluster/updateblob.go
+++ b/pkg/cluster/updateblob.go
@@ -3,6 +3,7 @@ package cluster
 import (
 	"bytes"
 	"encoding/json"
+	"sort"
 )
 
 type updateblob map[instanceName]hash
@@ -16,11 +17,17 @@ type vmInfo struct {
 }
 
 func (blob updateblob) MarshalJSON() ([]byte, error) {
+	instancenames := make([]instanceName, 0, len(blob))
+	for instancename := range blob {
+		instancenames = append(instancenames, instancename)
+	}
+	sort.Slice(instancenames, func(i, j int) bool { return instancenames[i] < instancenames[j] })
+
 	slice := make([]vmInfo, 0, len(blob))
-	for instancename, hash := range blob {
+	for _, instancename := range instancenames {
 		slice = append(slice, vmInfo{
 			InstanceName: instancename,
-			ScalesetHash: hash,
+			ScalesetHash: blob[instancename],
 		})
 	}
 

--- a/pkg/cluster/updateblob_test.go
+++ b/pkg/cluster/updateblob_test.go
@@ -1,0 +1,90 @@
+package cluster
+
+import (
+	"bytes"
+	"io/ioutil"
+	"reflect"
+	"strings"
+	"testing"
+
+	gomock "github.com/golang/mock/gomock"
+
+	"github.com/openshift/openshift-azure/pkg/util/mocks/mock_azureclient/mock_storage"
+)
+
+func TestReadUpdateBlob(t *testing.T) {
+	tests := []struct {
+		name    string
+		want    updateblob
+		wantErr string
+		blob    string
+	}{
+		{
+			name:    "empty",
+			wantErr: "unexpected end of JSON input",
+		},
+		{
+			name: "ok",
+			want: updateblob{
+				"ss-infra_0":   "45",
+				"ss-compute_0": "7x99=",
+			},
+			blob: `[{"instanceName": "ss-infra_0", "scalesetHash": "45"},{"instanceName":"ss-compute_0","scalesetHash":"7x99="}]`,
+		},
+	}
+	gmc := gomock.NewController(t)
+	for _, tt := range tests {
+		updateBlob := mock_storage.NewMockBlob(gmc)
+		data := ioutil.NopCloser(strings.NewReader(tt.blob))
+		updateBlob.EXPECT().Get(nil).Return(data, nil)
+		u := &simpleUpgrader{
+			updateBlob: updateBlob,
+		}
+
+		got, err := u.readUpdateBlob()
+		if (err != nil) != (len(tt.wantErr) > 0) {
+			t.Errorf("simpleUpgrader.readUpdateBlob() error = %v, wantErr %v", err, tt.wantErr)
+			return
+		}
+		if len(tt.wantErr) > 0 && !strings.Contains(err.Error(), tt.wantErr) {
+			t.Errorf("simpleUpgrader.readUpdateBlob() error = %v, wantErr %v", err, tt.wantErr)
+		}
+		if !reflect.DeepEqual(got, tt.want) && len(tt.wantErr) == 0 {
+			t.Errorf("simpleUpgrader.readUpdateBlob() = %v, want %v", got, tt.want)
+		}
+	}
+}
+
+func TestWriteUpdateBlob(t *testing.T) {
+	tests := []struct {
+		name    string
+		b       updateblob
+		wantErr string
+		blob    string
+	}{
+		{
+			name: "empty",
+			blob: "[]",
+		},
+		{
+			name: "valid",
+			b: updateblob{
+				"ss-infra_0":   "45",
+				"ss-compute_0": "7x99=",
+			},
+			blob: `[{"instanceName":"ss-infra_0","scalesetHash":"45"},{"instanceName":"ss-compute_0","scalesetHash":"7x99="}]`,
+		},
+	}
+	gmc := gomock.NewController(t)
+	for _, tt := range tests {
+		updateBlob := mock_storage.NewMockBlob(gmc)
+		updateBlob.EXPECT().CreateBlockBlobFromReader(bytes.NewReader([]byte(tt.blob)), nil)
+		u := &simpleUpgrader{
+			updateBlob: updateBlob,
+		}
+
+		if err := u.writeUpdateBlob(tt.b); (err != nil) != (tt.wantErr != "") {
+			t.Errorf("simpleUpgrader.writeUpdateBlob() error = %v, wantErr %v", err, tt.wantErr)
+		}
+	}
+}

--- a/pkg/cluster/updateblob_test.go
+++ b/pkg/cluster/updateblob_test.go
@@ -8,7 +8,7 @@ import (
 	"strings"
 	"testing"
 
-	gomock "github.com/golang/mock/gomock"
+	"github.com/golang/mock/gomock"
 
 	"github.com/openshift/openshift-azure/pkg/util/mocks/mock_azureclient/mock_storage"
 )
@@ -16,9 +16,9 @@ import (
 func TestReadUpdateBlob(t *testing.T) {
 	tests := []struct {
 		name    string
+		blob    string
 		want    updateblob
 		wantErr error
-		blob    string
 	}{
 		{
 			name:    "empty",
@@ -26,11 +26,11 @@ func TestReadUpdateBlob(t *testing.T) {
 		},
 		{
 			name: "ok",
+			blob: `[{"instanceName":"ss-infra_0","scalesetHash":"45"},{"instanceName":"ss-compute_0","scalesetHash":"7x99="}]`,
 			want: updateblob{
 				"ss-infra_0":   "45",
 				"ss-compute_0": "7x99=",
 			},
-			blob: `[{"instanceName": "ss-infra_0", "scalesetHash": "45"},{"instanceName":"ss-compute_0","scalesetHash":"7x99="}]`,
 		},
 	}
 	gmc := gomock.NewController(t)
@@ -59,32 +59,32 @@ func TestReadUpdateBlob(t *testing.T) {
 func TestWriteUpdateBlob(t *testing.T) {
 	tests := []struct {
 		name    string
-		b       updateblob
+		blob    updateblob
+		want    string
 		wantErr string
-		blob    string
 	}{
 		{
 			name: "empty",
-			blob: "[]",
+			want: "[]",
 		},
 		{
 			name: "valid",
-			b: updateblob{
+			blob: updateblob{
 				"ss-infra_0":   "45",
 				"ss-compute_0": "7x99=",
 			},
-			blob: `[{"instanceName":"ss-infra_0","scalesetHash":"45"},{"instanceName":"ss-compute_0","scalesetHash":"7x99="}]`,
+			want: `[{"instanceName":"ss-infra_0","scalesetHash":"45"},{"instanceName":"ss-compute_0","scalesetHash":"7x99="}]`,
 		},
 	}
 	gmc := gomock.NewController(t)
 	for _, tt := range tests {
 		updateBlob := mock_storage.NewMockBlob(gmc)
-		updateBlob.EXPECT().CreateBlockBlobFromReader(bytes.NewReader([]byte(tt.blob)), nil)
+		updateBlob.EXPECT().CreateBlockBlobFromReader(bytes.NewReader([]byte(tt.want)), nil)
 		u := &simpleUpgrader{
 			updateBlob: updateBlob,
 		}
 
-		if err := u.writeUpdateBlob(tt.b); (err != nil) != (tt.wantErr != "") {
+		if err := u.writeUpdateBlob(tt.blob); (err != nil) != (tt.wantErr != "") {
 			t.Errorf("simpleUpgrader.writeUpdateBlob() error = %v, wantErr %v", err, tt.wantErr)
 		}
 	}

--- a/pkg/cluster/updateblob_test.go
+++ b/pkg/cluster/updateblob_test.go
@@ -2,6 +2,7 @@ package cluster
 
 import (
 	"bytes"
+	"io"
 	"io/ioutil"
 	"reflect"
 	"strings"
@@ -16,12 +17,12 @@ func TestReadUpdateBlob(t *testing.T) {
 	tests := []struct {
 		name    string
 		want    updateblob
-		wantErr string
+		wantErr error
 		blob    string
 	}{
 		{
 			name:    "empty",
-			wantErr: "unexpected end of JSON input",
+			wantErr: io.EOF,
 		},
 		{
 			name: "ok",
@@ -42,14 +43,14 @@ func TestReadUpdateBlob(t *testing.T) {
 		}
 
 		got, err := u.readUpdateBlob()
-		if (err != nil) != (len(tt.wantErr) > 0) {
+		if (err != nil) != (tt.wantErr != nil) {
 			t.Errorf("simpleUpgrader.readUpdateBlob() error = %v, wantErr %v", err, tt.wantErr)
 			return
 		}
-		if len(tt.wantErr) > 0 && !strings.Contains(err.Error(), tt.wantErr) {
+		if tt.wantErr != nil && err != tt.wantErr {
 			t.Errorf("simpleUpgrader.readUpdateBlob() error = %v, wantErr %v", err, tt.wantErr)
 		}
-		if !reflect.DeepEqual(got, tt.want) && len(tt.wantErr) == 0 {
+		if tt.wantErr == nil && !reflect.DeepEqual(got, tt.want) {
 			t.Errorf("simpleUpgrader.readUpdateBlob() = %v, want %v", got, tt.want)
 		}
 	}

--- a/pkg/cluster/updateblob_test.go
+++ b/pkg/cluster/updateblob_test.go
@@ -26,7 +26,7 @@ func TestReadUpdateBlob(t *testing.T) {
 		},
 		{
 			name: "ok",
-			blob: `[{"instanceName":"ss-infra_0","scalesetHash":"45"},{"instanceName":"ss-compute_0","scalesetHash":"7x99="}]`,
+			blob: `[{"instanceName":"ss-compute_0","scalesetHash":"7x99="},{"instanceName":"ss-infra_0","scalesetHash":"45"}]`,
 			want: updateblob{
 				"ss-infra_0":   "45",
 				"ss-compute_0": "7x99=",
@@ -73,7 +73,7 @@ func TestWriteUpdateBlob(t *testing.T) {
 				"ss-infra_0":   "45",
 				"ss-compute_0": "7x99=",
 			},
-			want: `[{"instanceName":"ss-infra_0","scalesetHash":"45"},{"instanceName":"ss-compute_0","scalesetHash":"7x99="}]`,
+			want: `[{"instanceName":"ss-compute_0","scalesetHash":"7x99="},{"instanceName":"ss-infra_0","scalesetHash":"45"}]`,
 		},
 	}
 	gmc := gomock.NewController(t)

